primitives/src/signatures/bls.rs

```diff
@@ -1 +1,165 @@
 //! Placeholder for BLS signature.
+
+use super::SignatureScheme;
+use crate::{constants::CS_ID_BLS_SIG_NAIVE, errors::PrimitivesError, hash_to_group::HashToGroup};
+use ark_ec::{
+    bls12::{Bls12, Bls12Parameters},
+    short_weierstrass_jacobian::GroupAffine,
+    AffineCurve, ModelParameters, ProjectiveCurve,
+};
+use ark_ff::{Fp12, PrimeField};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, SerializationError, Write};
+use ark_std::{
+    io::Read,
+    ops::Neg,
+    rand::{CryptoRng, RngCore},
+    string::ToString,
+    One, UniformRand,
+};
+use core::marker::PhantomData;
+use jf_utils::{multi_pairing, tagged_blob};
+
+/// BLS signature scheme.
+/// Optimized for signature size, i.e.: PK in G2 and sig in G1
+pub struct BLSSignatureScheme<P: Bls12Parameters> {
```

Comment:
`P: Bls12Parameters`: Does this mean that we'll only use BLS curves for BLS signatures? That is, we won't use other pairing-friendly curves?

Reply:
Yeah. I tried to make it generic over `E: PairingEngine` but `PairingEngine` does not expose `ModelParameters` which is needed for hashes, etc.

Reply:
I see, sounds good for me.